tool/python/singa/layer.py

```diff
@@ -41,13 +45,110 @@ def __init__(self, **kwargs):
         '''
 
         self.layer = Message('Layer', **kwargs).proto
-        # required
+        # required field
         if not 'name' in kwargs:
             setval(self.layer, name=generate_name('layer', 1))
 
-        # srclayers are set in Model.build()
+        # layer connectivity is set in Model.build()
         self.is_datalayer = False
 
+        self.singalayer = None
+        self.srclayers = [] 
+
+    def ComputeFeature(self, *srclys):
+        #print ' >> CompFeature()', self.layer.name, '# of srcs:', len(srclys)
+
+        # create singa::Layer and store srclayers 
+        if self.singalayer == None:
+            self.singalayer = SingaLayer.CreateLayer(self.layer.SerializeToString())  
+            self.singaSrclayerVector = layerVector(len(srclys)) 
+            for i in range(len(srclys)):
+                self.srclayers.append(srclys[i])
+                self.singaSrclayerVector[i] = srclys[i].get_singalayer()
+
+            #print ' >> Setup', self.layer.name
+            SingaLayer.SetupLayer(self.singalayer, self.layer.SerializeToString(), self.singaSrclayerVector)
+
+        self.singalayer.ComputeFeature(1, self.singaSrclayerVector)
+
+
+    def ComputeGradient(self, step, upd=None):
+        #print ' >> CompGradient()', self.layer.name, '# of srcs', len(self.srclayers)
+
+        # create singa::Updater
+        assert upd != None, 'required Updater (see model.py)' 
+        if Layer.singaupdater == None:
+            Layer.singaupdater = SingaUpdater.CreateUpdater(upd.proto.SerializeToString()) 
+
+        # call ComputeGradient of Singa
+        self.singalayer.ComputeGradient(1, self.singaSrclayerVector)
+        SingaUpdater.UpdateParams(Layer.singaupdater, self.singalayer, step)
+
+        # recursively call ComputeGradient of srclayers
+        for sly in self.srclayers:
```

Comment:
If a layer has multiple children, the layer's computeGradient method will be called multiple times.